File: frontend/src/services/projects.ts
import api from "./client";
import type { Project } from "../types";

export async function fetchProjects(): Promise<Project[]> {
  const { data } = await api.get<Project[]>("/projects/");
  return data;
}

export async function createProject(data: { name: string; color: string; description?: string }): Promise<Project> {
  const { data: created } = await api.post<Project>("/projects/", data);
  return created;
}

export async function updateProject(id: string, data: { name?: string; color?: string; description?: string }): Promise<Project> {
  const { data: updated } = await api.put<Project>(`/projects/${id}`, data);
  return updated;
}

export async function deleteProject(id: string): Promise<void> {
  await api.delete(`/projects/${id}`);
}

File: api/app/api/routes/projects.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.repositories.project_repository import ProjectRepository
from app.schemas.project import ProjectCreate, ProjectResponse, ProjectUpdate
from app.services.project_service import ProjectService

router = APIRouter(prefix="/api/projects", tags=["projects"])


def get_service(db: Session = Depends(get_db)) -> ProjectService:
    return ProjectService(ProjectRepository(db))


@router.get("/", response_model=list[ProjectResponse])
def list_projects(service: ProjectService = Depends(get_service)):
    return service.list_all()


@router.post("/", response_model=ProjectResponse, status_code=status.HTTP_201_CREATED)
def create_project(data: ProjectCreate, service: ProjectService = Depends(get_service)):
    return service.create(data)


@router.get("/{project_id}", response_model=ProjectResponse)
def get_project(project_id: str, service: ProjectService = Depends(get_service)):
    return service.get_by_id(project_id)


@router.put("/{project_id}", response_model=ProjectResponse)
def update_project(project_id: str, data: ProjectUpdate, service: ProjectService = Depends(get_service)):
    return service.update(project_id, data)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(project_id: str, service: ProjectService = Depends(get_service)):
    service.delete(project_id)
